src/generate-grpc-web.ts

```diff
@@ -148,33 +157,50 @@ function methodDescName(serviceDesc: ServiceDescriptorProto, methodDesc: MethodD
 export function addGrpcWebMisc(options: Options, _file: FileSpec): FileSpec {
   let file = _file;
   file = file.addCode(
-    CodeBlock.empty().addStatement('type UnaryMethodDefinitionish = grpc.UnaryMethodDefinition<any, any>')
+    CodeBlock.empty()
+      .addStatement('import UnaryMethodDefinition = grpc.UnaryMethodDefinition')
+      .addStatement(
+        'interface UnaryMethodDefinitionishR extends UnaryMethodDefinition<any, any> { requestStream: any; responseStream: any; }'
+      )
+      .addStatement('type UnaryMethodDefinitionish = UnaryMethodDefinitionishR')
   );
-  file = file.addInterface(generateGrpcWebRpcType());
-  file = file.addClass(generateGrpcWebImpl());
+  file = file.addInterface(generateGrpcWebRpcType(options.returnObservable));
+  file = file.addClass(options.returnObservable ? generateGrpcWebImplObservable() : generateGrpcWebImplPromise());
   return file;
 }
 
-/** Makes an `Rpc` interface to decouple from the low-level grpc-web `grpc.unary`/etc. methods. */
-function generateGrpcWebRpcType(): InterfaceSpec {
+/** Makes an `Rpc` interface to decouple from the low-level grpc-web `grpc.invoke and grpc.unary`/etc. methods. */
+function generateGrpcWebRpcType(returnObservable: boolean): InterfaceSpec {
   let rpc = InterfaceSpec.create('Rpc');
-  let fn = FunctionSpec.create('unary');
+  let fnU = FunctionSpec.create('unary');
+  let fnI = FunctionSpec.create('invoke');
   const t = TypeNames.typeVariable('T', TypeNames.bound('UnaryMethodDefinitionish'));
-  fn = fn
+  fnU = fnU
     .addTypeVariable(t)
     .addParameter('methodDesc', t)
     .addParameter('request', TypeNames.ANY)
     .addParameter('metadata', TypeNames.unionType(TypeNames.anyType('grpc.Metadata'), TypeNames.UNDEFINED))
-    .returns(TypeNames.PROMISE.param(TypeNames.ANY));
-  rpc = rpc.addFunction(fn);
+    .returns(
+      returnObservable
+        ? TypeNames.anyType('Observable@rxjs').param(TypeNames.ANY)
+        : TypeNames.PROMISE.param(TypeNames.ANY)
+    );
+  fnI = fnI
+    .addTypeVariable(t)
+    .addParameter('methodDesc', t)
+    .addParameter('request', TypeNames.ANY)
+    .addParameter('metadata', TypeNames.unionType(TypeNames.anyType('grpc.Metadata'), TypeNames.UNDEFINED))
+    .returns(TypeNames.anyType('Observable@rxjs').param(TypeNames.ANY));
+  rpc = rpc.addFunction(fnU).addFunction(fnI);
   return rpc;
 }
 
 /** Implements the `Rpc` interface by making calls using the `grpc.unary` method. */
-function generateGrpcWebImpl(): ClassSpec {
+function generateGrpcWebImplPromise(): ClassSpec {
   const maybeMetadata = TypeNames.unionType(TypeNames.anyType('grpc.Metadata'), TypeNames.UNDEFINED);
   const optionsParam = TypeNames.anonymousType(
-    ['transport?', TypeNames.anyType('grpc.TransportFactory')],
+    ['unaryTransport?', TypeNames.anyType('grpc.TransportFactory')],
```

Comment:
@PhilipMantrov out of curiosity, do the `unaryTransport` and `invokeTransport` have to be different? 

...also, this is a per-request override anyway, so I think it'd be simpler to do just `transport` and have the user not care about if its supposed to be `unary` or `invoke` under the hood.

I'm planning on changing this to just `transport` like it was before, but let me know if there is some rationale here that I don't know about.

Reply:
Often they use websocket transport for streams, and standard transport for unary. We should be able to use different transports for unary and invoke methods.

Reply:
Ah okay. Just for my own understanding, could the websocket transport be used for both unary and invoke, if a user wanted to do that?

Fwiw I also missed that this is a breaking change to the api. I'm thinking about keeping the old transport name for the unary and also invoke by default, and then adding a new 'streamingTransport', that would be for the invoke usage (but invoke would use the 'transport' if 'streamingTranspory' isn't set).

Kind of minor, but I like steamingTransport as a less "implementation details" name instead of the user knowing about invoke vs unary.

Sound okay?

Reply:
>Just for my own understanding, could the websocket transport be used for both unary and invoke, if a user wanted to do that?

Yes he can, but don't forget about active websocket connection limitations for modern browsers (6 per page and 255 per browser).